### PR TITLE
added lv_obj_move_to_index(obj, index), renamed lv_obj_get_child_id(obj) to lv_obj_get_index(obj)

### PR DESCRIPTION
**Pull request by Karijn**
- added lv_obj_move_to_index(obj, index).
- renamed lv_obj_get_child_id(obj) to lv_obj_get_index(obj).

### Description of the feature or fix

A clear and concise description of what the bug or new feature is.

### Checkpoints
- [x] Follow the [styling guide](https://github.com/lvgl/lvgl/blob/master/docs/CODING_STYLE.md)
- [x] Update [CHANGELOG.md](https://github.com/lvgl/lvgl/blob/master/docs/CHANGELOG.md)
- [x] Update the documentation

### CODE REVIEW

**Comment by Karijn** [comment]
Hi, the inline function did give an error during the MP Build, other options i tried did not work either.

File: src/core/lv_obj_tree.h [hunk 1]
```diff
@@ -166,11 +174,20 @@ uint32_t lv_obj_get_child_cnt(const struct _lv_obj_t * obj);
...
+
+/**
+ * DEPRECATED FUNCTION, please use uint32_t lv_obj_get_index(obj);
+ * Get the index of a child.
+ * @param obj       pointer to an object
  * @return          the child index of the object.
  *                  E.g. 0: the oldest (firstly created child)
  */
-uint32_t lv_obj_get_child_id(const struct _lv_obj_t * obj);
+uint32_t lv_obj_get_child_id(const struct _lv_obj_t* obj);
```
Comment (embeddedt) [inline src/core/lv_obj_tree.h]: Making it `static inline` in this header instead of `lv_api_map.h` should work.

Comment (Karijn) [reply src/core/lv_obj_tree.h]: that did work, thanks

File: src/lv_api_map.h [hunk 2]
```diff
@@ -45,13 +45,11 @@ static inline LV_ATTRIBUTE_TIMER_HANDLER uint32_t lv_task_handler(void)
   * DEPRECATED FUNCTIONS
   **********************/
 
-/* function is inline in lv_obj_tree.h, otherwise Micropython will not build
-inline uint32_t lv_obj_get_child_id(const struct _lv_obj_t* obj)
+static inline uint32_t lv_obj_get_child_id(const struct _lv_obj_t* obj)
 {
     LV_LOG_USER("lv_obj_get_child_id(obj) is deprecated, please use lv_obj_get_index(obj).");
```
Comment (kisvegabor) [inline src/lv_api_map.h]: `LV_LOG_USER` is for logs sent by the user. I'd use `LV_LOG_WARN` here.

File: src/core/lv_obj_tree.c [hunk 3]
```diff
@@ -249,23 +287,24 @@ void lv_obj_swap(lv_obj_t * obj1, lv_obj_t * obj2)
...
+
+    if(parent != parent2)
+    {
         lv_obj_invalidate(parent2);
     }
-
     lv_group_swap_obj(obj1, obj2);
 }
 
 void lv_obj_move_up(lv_obj_t * obj)
```
Comment (kisvegabor) [inline src/core/lv_obj_tree.c]: Shouldn't  `lv_obj_move_up/down` be removed?

File: src/core/lv_obj_tree.h [hunk 4]
```diff
@@ -123,6 +123,14 @@ void lv_obj_move_up(struct _lv_obj_t* obj);
  */
 void lv_obj_move_down(struct _lv_obj_t* obj);
 
+/**
+ * moves the object to the given index in its parent.
+ * When used in listboxes, it can be used to sort the listbox items.
+ * @param obj  pointer to the object to be moved.
+ * @param index  new index in parent.
```
Comment (kisvegabor) [inline src/core/lv_obj_tree.h:100]: I'd be great to have some examples for specific cases. E.g.
```doxygen
@note to move to the foreground: lv_obj_move_to_index(obj, 0)
@note to move forward (up): lv_obj_move_to_index(obj, lv_obj_get_index(obj) - 1)
```

**Comment by kisvegabor** [comment]
Thanks you, I had a few comments.

File: src/core/lv_obj_tree.c (hunk 3 above)
Comment (Karijn) [reply src/core/lv_obj_tree.c]: I was not sure about removing these functions, the nice thing about them is that they will take care of calculating and testing the new position.
but i can remove them.

Comment (kisvegabor) [reply src/core/lv_obj_tree.c]: If index is too small or too large lv_obj_move_to_index will return too.

Comment (Karijn) [reply src/core/lv_obj_tree.c]: done

Comment (kisvegabor) [reply src/core/lv_obj_tree.c]: Thanks!

File: src/lv_api_map.h (hunk 2 above)
Comment (Karijn) [reply src/lv_api_map.h]: done

File: src/core/lv_obj_tree.h (hunk 4 above)
Comment (Karijn) [reply src/core/lv_obj_tree.h:100]: done

**Comment by Karijn** [comment]
i think i solved all comments (or did i miss one)?

**Comment by kisvegabor** [comment]
Thanks you, only one question remained  for me:
I think `lv_obj_move_background/foregrund` should be only `static inline` wrappers to `lv_obj_move_to_index()`.

What do you think?

**Comment by Karijn** [comment]
> Thanks you, only one question remained for me:
> I think `lv_obj_move_background/foregrund` should be only `static inline` wrappers to `lv_obj_move_to_index()`.
> 
> What do you think?

No reason nof to, i think. I will create the inlines tonight.

**Comment by Karijn** [comment]
redefined lv_obj_move_foreground(obj) and lv_obj_move_background(obj) as inline functions now calling lv_obj_move_to_index(obj, index).

@kisvegabor I have only one concern with this change: will these functions still be picked up by MicroPython?

**Comment by kisvegabor** [comment]
>  I have only one concern with this change: will these functions still be picked up by MicroPython?

Yes, the function in `lv_api_map.h` can be seen in MicroPython.

---

All looks good to me. I'm very happy with the result. Now we have only a simple but very flexible function for child order manipulation. 

Thank you very much!

**Comment by Karijn** [comment]
Hi,

I see that the new functions are in the latest lv_micropython version. But how can i enable them in micropython (I don't find them in the lv_mpy_example.c file)

**Comment by amirgon** [comment]
@Karijn I can see you updated the C examples.
Are you planning to update the corresponding Micropython examples?

**Comment by Karijn** [comment]
@amirgon, yes definitely. As soon as i have the new and changed functions working in MP,  I'll do a PR for the samples and documentation for MicroPython. (actually, i needed those functions in MP in the first place :-))

**Comment by amirgon** [comment]
> As soon as i have the new and changed functions working in MP

They are working for me on MP.  
Any specific problem I can help you with?

**Comment by Karijn** [comment]
[PR for MicroPython sample is here](https://github.com/lvgl/lvgl/pull/2545)

**Pull request merged**